### PR TITLE
Using ORTModule to wrap a evaluation model should not change the mode

### PR DESCRIPTION
**Pull request by guyang3532**
To fix the bug:
When using ORTModule wrap model and do evaluation, the mode of model will become training after ORT wrapping.

### CODE REVIEW

**Comment by guyang3532** [comment]
@pengwa

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 1]
```diff
@@ -5334,3 +5334,12 @@ def run_step(model, x):
     _test_helpers.assert_values_are_close(pt_prediction, ort_prediction)
     _test_helpers.assert_values_are_close(pt_loss, ort_loss)
     _test_helpers.assert_values_are_close(pt_x.grad, ort_x.grad)
+
+
+def test_eval_model_mod():
```
Comment (AdamLouly) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:5339]: typo here: test_eval_model_mod , should be test_eval_model_mode

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py [hunk 2]
```diff
@@ -102,6 +102,8 @@ def __init__(self, module, debug_options=None):
         # Assign self._is_initialized to True after all the ORTModule class attributes have been assigned
         # else, they will be assigned to self._torch_module.original_module instead.
         self._is_initialized = True
+        if not module.training:
+            self.eval()
```
Comment (AdamLouly) [inline orttraining/orttraining/python/training/ortmodule/ortmodule.py:106]: let's try to keep, `self._is_initialized = True` as the last step on the constructor.

**Review by AdamLouly — CHANGES_REQUESTED** [review]

**Comment by baijumeswani** [comment]
we probably need to update the test `test_bert_inputs_with_dynamic_shape` to change the mode to be training based on https://github.com/huggingface/transformers/blob/v4.21.2/src/transformers/modeling_utils.py#L1588

**Comment by thiagocrepaldi** [comment]
Unless that changed, `ORTModule` should only support training ONNX graphs, that is why it changes the mode to training before exporting.

I believe open vino integration (a similar frontend to ORTModule) should support inference, though.

**Comment by baijumeswani** [comment]
@thiagocrepaldi We still want to use ORTModule for training scenarios, that has not changed.

Having said that, I believe ORTModule should inherit the training mode from the original pytorch model. So, if the original model is instantiated and is in eval mode, ORTModule should inherit that mode. Otherwise, there is an inconsistency between the intent and ORTModule implementation.

For example, if the user passes in a model in eval mode, ORTModule will by default set its own `super().training` to `True`, but the underlying objects `TrainingManager` and `InferenceManager` will be constructed based on the mode of the original pytorch model. So, this is inconsistent and needs to be fixed.

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi We still want to use ORTModule for training scenarios, that has not changed.
> 
> Having said that, I believe ORTModule should inherit the training mode from the original pytorch model. So, if the original model is instantiated and is in eval mode, ORTModule should inherit that mode. Otherwise, there is an inconsistency between the intent and ORTModule implementation.
> 
> For example, if the user passes in a model in eval mode, ORTModule will by default set its own `super().training` to `True`, but the underlying objects `TrainingManager` and `InferenceManager` will be constructed based on the mode of the original pytorch model. So, this is inconsistent and needs to be fixed.

If the model is left in eval(), the dropout nodes will never be added to the static onnx graph, and therefore, it wont be able to be used for training later on

**Comment by baijumeswani** [comment]
> If the model is left in eval(), the dropout nodes will never be added to the static onnx graph, and therefore, it wont be able to be used for training later on
>
> For your idea to work, you would have to keep track of traning/eval model and trigger an onnx re-export when the mode changes... for a big model, this is a couple minutes worth of delay. the perf may not be negligible when compared to pytorch perf

We already keep track of the mode in `ORTModule`. The model is exported once per mode. So, the training model (with dropout) is exported and stored inside the `TrainingManager` and the eval model is exported and stored inside the `InferenceManager`.

**Comment by pengwa** [comment]
> > If the model is left in eval(), the dropout nodes will never be added to the static onnx graph, and therefore, it wont be able to be used for training later on
> > For your idea to work, you would have to keep track of traning/eval model and trigger an onnx re-export when the mode changes... for a big model, this is a couple minutes worth of delay. the perf may not be negligible when compared to pytorch perf
> 
> We already keep track of the mode in `ORTModule`. The model is exported once per mode. So, the training model (with dropout) is exported and stored inside the `TrainingManager` and the eval model is exported and stored inside the `InferenceManager`.

This is also my understanding.  TorchModuleORT's GraphExecutionManagerFactory helps hide the eval/train when calling model forward already, isn't it the reason we call it a factory? @thiagocrepaldi .

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 1 above)
Comment (guyang3532) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: thx, done

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py (hunk 2 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/ortmodule.py]: done

[the remark this answers is not in the document]
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/ortmodule.py]: done, thx

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 3]
```diff
@@ -2014,6 +2014,7 @@ def test_bert_inputs_with_dynamic_shape():
     pt_model = _get_bert_for_sequence_classification_model(
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2014]:
```suggestion
    pt_model = _get_bert_for_sequence_classification_model(is_training=True,
```
Having that on the API would be better for reuse. Maybe setting default to `False` as it is the official behavior of `BertForSequenceClassification.from_pretrained()`

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -5334,3 +5335,12 @@ def run_step(model, x):
...
     _test_helpers.assert_values_are_close(pt_x.grad, ort_x.grad)
+
+
+def test_eval_model_mode():
+    device = "cuda"
+    d_in, h_size, d_out = 2, 2, 2
+    model = NeuralNetSinglePositionalArgument(d_in, h_size, d_out).to(device)
+    model.eval()
+    ort_model = ORTModule(model)
+    assert ort_model.training is False
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:5346]:
```suggestion
    assert ort_model.training is False
    assert ort_model._torch_module.training is False
```
Just to make sure the ORTModule instance (user facing torch.nn.Module) matches the internal one (original module before wrapping)

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi We still want to use ORTModule for training scenarios, that has not changed.
> 
> Having said that, I believe ORTModule should inherit the training mode from the original pytorch model. So, if the original model is instantiated and is in eval mode, ORTModule should inherit that mode. Otherwise, there is an inconsistency between the intent and ORTModule implementation.
> 
> For example, if the user passes in a model in eval mode, ORTModule will by default set its own `super().training` to `True`, but the underlying objects `TrainingManager` and `InferenceManager` will be constructed based on the mode of the original pytorch model. So, this is inconsistent and needs to be fixed.

It would be great having this info on the PR description. It lacks context for future tracking

**Comment by thiagocrepaldi** [comment]
Oh my, I almost forgot about the execution managers, you are right.

Both inference and training managers are instantiated during `ORTMOdule.__init__` and one of them is chosen during `ORTModule.forward`, based on training mode was set during `ORTMModule.__init__` time

If the training mode changes after that, maybe this change will not be captured, say after a training loop we call a test loop. I believe this would cause the training model to be called, even with this change.

I suggested a change on the unit test to check this scenario (call model.eval() and model.train() before first forward step and later call both a couple times after export)

**Comment by guyang3532** [comment]
> Oh my, I almost forgot about the execution managers, you are right.
> 
> Both inference and training managers are instantiated during `ORTMOdule.__init__` and one of them is chosen during `ORTModule.forward`, based on training mode was set during `ORTMModule.__init__` time
> 
> If the training mode changes after that, maybe this change will not be captured, say after a training loop we call a test loop. I believe this would cause the training model to be called, even with this change.
> 
> I suggested a change on the unit test to check this scenario (call model.eval() and model.train() before first forward step and later call both a couple times after export)

@thiagocrepaldi Do you mean example code like:
               model.train()
               model.forward()
               model.eval()
               model.forward()
?? Why the mode change will not be captured in the forward if we call model.train() or model.eval() before the forward?

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py [hunk 5]
```diff
@@ -101,6 +101,7 @@ def __init__(self, module, debug_options=None):
         # Finally, ORTModule initialization is complete.
         # Assign self._is_initialized to True after all the ORTModule class attributes have been assigned
         # else, they will be assigned to self._torch_module.original_module instead.
+        self.train(module.training)
```
[the remark this answers is not in the document]
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/ortmodule.py]: thx, done

**Review by baijumeswani — APPROVED** [review]

**Comment by guyang3532** [comment]
@thiagocrepaldi @AdamLouly

**Comment by guyang3532** [comment]
> > > Oh my, I almost forgot about the execution managers, you are right.
> > > Both inference and training managers are instantiated during `ORTMOdule.__init__` and one of them is chosen during `ORTModule.forward`, based on training mode was set during `ORTMModule.__init__` time
> > > If the training mode changes after that, maybe this change will not be captured, say after a training loop we call a test loop. I believe this would cause the training model to be called, even with this change.
> > > I suggested a change on the unit test to check this scenario (call model.eval() and model.train() before first forward step and later call both a couple times after export)
> > 
> > 
> > @thiagocrepaldi Do you mean example code like: model.train() model.forward() model.eval() model.forward() ?? Why the mode change will not be captured in the forward if we call model.train() or model.eval() before the forward?
> 
> Yes, something like this, or
> 
> ```python
> for initial_mode in (True, False):
>     model = copy.deepcopy(ORTModule(MyModel(...)))
>     input = copy.deepcopy(*args, **kwargs)
>     model.train(initial_mode)
> 
>     for step in range(2):
>         for new_mode in (True, False):
>             model.train(new_mode)
>             output = model(input)
>             assert is_dropout_node_or_any_training_specific_node_in_the_graph() == new_mode #go through the graph looking for such nodes)
>             assert ort_model.training == new_mode
>             assert ort_model._torch_module.training == new_mode
>             assert ort_model._torch_module._flattened_module.training == new_mode
> ```
> 
> This is just to make sure that all models (public ORTModule and its hidden ones) have the correct training mode after each forward step, especially after the training mode is flipped. I am not 100% that happens, as we've always focused only on training for the tests. Maybe we do export both training and inference, but always use the first seen training mode

done, thx.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 6]
```diff
@@ -414,6 +415,8 @@ def _get_bert_for_sequence_classification_model(
         config=config,
     ).to(device)
 
+    if is_training:
+        model.train()
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:419]:
```suggestion
        model.train()
    else:
        model.eval()
```
To covert the case where `is_training is False and `model.is_training is True`

Comment (guyang3532) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:419]: done. thx

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by guyang3532** [comment]
@AdamLouly

**Review by AdamLouly — APPROVED** [review]
LGTM
